### PR TITLE
feat: support non-zero std types

### PR DESCRIPTION
**Pull request by hrls**

### CODE REVIEW

File: borsh/src/ser/mod.rs [hunk 1]
```diff
@@ -94,6 +94,31 @@ impl_for_integer!(u32);
 impl_for_integer!(u64);
 impl_for_integer!(u128);
 
+mod non_zero {
+    use super::*;
+    use std::num::*;
```
Comment (frol) [inline borsh/src/ser/mod.rs:99]: To support no-std context, we should use `num` from `core` instead of `std`:

```suggestion
    use core::num::*;
```

File: borsh/src/de/mod.rs [hunk 2]
```diff
@@ -143,6 +143,73 @@ impl BorshDeserialize for usize {
...
+                        if buf.len() < primitive_size {
+                            return Err(Error::new(
+                                ErrorKind::InvalidInput,
+                                ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+                            ));
+                        }
+                        let primitive = <$type as NonZero>::Primitive::from_le_bytes(
+                            buf[..primitive_size].try_into().unwrap(),
+                        );
+                        let non_zero = <$type>::new(primitive).unwrap();
```
Comment (frol) [inline borsh/src/de/mod.rs:197]: This will crash an application if 0 gets deserialized, so instead of `unwrap` we should map it to Result and return with an error.

Please, add a test for this case.

**Review by frol — CHANGES_REQUESTED** [review]
@hrls Thank you for your contribution! Please, address review comments and add more tests :pray:

**Review by ailisp — COMMENTED** [review]
Looks good! Just need to address @frol on not panic

File: borsh/src/de/mod.rs [hunk 3]
```diff
@@ -143,6 +143,73 @@ impl BorshDeserialize for usize {
...
+        NonZeroU128 => u128,
+
+        NonZeroI8   => i8,
+        NonZeroI16  => i16,
+        NonZeroI32  => i32,
+        NonZeroI64  => i64,
+        NonZeroI128 => i128,
+
+        NonZeroUsize => usize,
+        NonZeroIsize => isize,
```
Comment (matklad) [inline borsh/src/de/mod.rs:178]: If I am reading code correctly, for serializing `92usize` this'll use 8 bytes on 64 bit system and 4 bytes on 32 bit system. This is wrong, serialization format should be independent of the host architecture. See how the plain `usize` is handled 

https://github.com/near/borsh-rs/blob/735686d9345202e8b8ee1f35eab52d7db6bb82c5/borsh/src/ser/mod.rs#L76-L81

To avoid this (and potentially other) gotchas, I suggest to change the implementation to *delegate* to impls of primiteve types, rather than re-implement the logic. That is, BorshDeserialize for NonZeroU32 should  call `<u32 as BorshDeserialize>::desrialize`, and then do just a single additional check for zero.

File: borsh/src/de/mod.rs [hunk 4]
```diff
@@ -144,6 +145,92 @@ impl BorshDeserialize for usize {
...
+    }
+
+    macro_rules! impl_for_non_zero_integer {
+        ($type:ty) => {
+            impl BorshDeserialize for $type {
+                #[inline]
+                fn deserialize(buf: &mut &[u8]) -> Result<Self> {
+                    let value: <$type as NonZero>::Primitive = <_>::deserialize(buf)?;
+                    <$type>::new(value)
+                        .ok_or_else(|| Error::new(ErrorKind::InvalidInput, ERROR_NON_ZERO_EXPECTED))
```
Comment (hrls) [inline borsh/src/de/mod.rs:189]: I'm not sure `InvalidInput` or `InvalidData` error kind should be used there.

File: borsh/src/de/mod.rs (hunk 3 above)
Comment (hrls) [reply borsh/src/de/mod.rs]: Done, probably

File: borsh/src/de/mod.rs (hunk 2 above)
Comment (hrls) [reply borsh/src/de/mod.rs]: Please review error kind and new error message.

File: borsh/src/de/mod.rs (hunk 4 above)
Comment (frol) [reply borsh/src/de/mod.rs:189]: I feel `InvalidInput` is fine, we use it in `bool` deserialization as well, but thanks for raising the concern :+1:

**Review by frol — APPROVED** [review]
LGTM

**Comment by frol** [comment]
@matklad May I ask you to take another look at it?

File: borsh/src/de/mod.rs (hunk 4 above)
Comment (matklad) [reply borsh/src/de/mod.rs:189]: Good point: we should be consistent first and foremost,  but indeed it'd be cool to fix separaetely (definitelly not in this PR : https://github.com/near/borsh-rs/issues/48)

File: borsh/src/de/mod.rs [hunk 5]
```diff
@@ -144,6 +145,92 @@ impl BorshDeserialize for usize {
...
+        NonZeroU64  => u64,
+        NonZeroU128 => u128,
+
+        NonZeroI8   => i8,
+        NonZeroI16  => i16,
+        NonZeroI32  => i32,
+        NonZeroI64  => i64,
+        NonZeroI128 => i128,
+
+        NonZeroUsize => usize,
```
Comment (matklad) [inline borsh/src/de/mod.rs:179]:
```suggestion
        NonZeroUsize => usize,
        NonZeroIsize => isize,
```

**Review by matklad — COMMENTED** [review]
Let's add a positive tests here (ie, the one which does't try to serialize zero). In `test_simple_structs.rs`, let's add an `nz: NonZeroU16` field.

File: borsh/src/de/mod.rs [hunk 6]
```diff
@@ -144,6 +145,92 @@ impl BorshDeserialize for usize {
     }
 }
 
+mod non_zero {
+    use super::*;
+    use core::num::*;
+
+    trait NonZero {
+        type Primitive;
+    }
```
Comment (austinabell) [inline borsh/src/de/mod.rs:154]: This trait doesn't seem necessary, since you're manually including the internal type in the macro, you can just generate the impl without it. I think this trait just adds complexity imo

**Comment by matklad** [comment]
Let's close this in favor of https://github.com/near/borsh-rs/pull/66 -- #66 manages to achieve the desired result with less code.
